[2단계 - 자동차 경주 리팩토링] 홍동건(쿤) 미션 제출합니다.

Hongdonggeon opened the pull request:
안녕하세요 로운님!! 쿤입니다.
리뷰를 너무 친절하게 달아주셔서 너무 많이 배울 수 있었던 것 같습니다!! 감사합니다 😭
피드백을 주시면서 첨부해주신 자료들 또한 너무 설명이 잘 되어 있어서 몇번이고 정독했던 것 같아요!!
인터페이스를 생성하여 테스트를 위한 구현체, 프로덕션 코드를 위한 구현체를 따로 두는 방법은 너무 신기했던 것 같아요
저번 리뷰에 랜덤값을 잘 만들어주는지의 테스트 코드는 따로 작성할 필요가 없는지에 대해 물어봤었는데, 잘 생각해보니 랜덤값이 4 이상일 때 잘 전진하는지와 4 미만일 때 전진하지 않는지만 검증을 해주면 되겠더라구요!!  그러기 위해, 4와 3을 반환해주는 구현체를 만든 것이고요!! 
새로운 개념들에 대해서 공부하면서 너무 유익한 일주일이었던 것 같습니다!! 다시 한번 감사드립니다!!

## 수정한 부분
### NumberGenerator 인터페이스 생성
- 프로덕션 코드에서는 RandomNumberGenrator라는 랜덤값을 반환해주는 구현체 사용
- 테스트 코드에서는 MovableNumberGenerator, NonMovableNumberGenerator라는 4와 3을 반환해주는 테스트 코드를 위한 구현체 사용
- Car의 move() 메서드 구조 변경 -> move(NumberGenerator numberGenerator)

### 일급 컬렉션 사용
- 일급 컬렉션에 메서드를 만들어가면서 최대한 사용
- 일급 컬렉션에서 List<Car>와 position을 비교해주는 메서드를 만들어주면서 외부에서 사용
- Set에서 List로 변경, Set은 순서 보장이 되지 않기 때문에 입력한 순서대로 결과 출력이 안되었기 때문에 List로 변경했습니다.

### Controller와 Game 클래스 분리
- Controller에는 View와 비즈니스 로직을 조립을 해주고, 전체 플로우를 관리하는 기능만 부여 (상태를 변하게 하는 기능들을 제외시켰습니다.)
- Game에는 Cars를 움직이게 해주는 기능, 우승자를 판정하는 기능 (게임에 핵심적인 로직 담당)

YJGwon committed:
test: String API에 대한 학습 Test class 추가

junitstudy/StringTest.java 추가
   - split_구분자_있을때(), split_구분자_없을때(): 구분자 단위 문자열 분리 test, containsExactly()는 배열의 모든 값이 일치해야 통과
   - subString_length_소괄호_제거(): length()를 이용해서 input값의 길이가 달라져도 항상 맨 앞, 맨 끝 문자를 제거
   - subString_index_소괄호_제거(): indexOf()를 이용해서 input값의 형식이 달라져도 항상 여는 괄호와 닫는 괄호를 제거
   - charAt_특정위치_문자_반환(): charAt()의 return 값을 String.valueOf()로 형변환
   - indexOutOfBoundsException_발생(): 발생한 Exception의 type이 IndexOutOfBoundsException인지 검사
   - exception_message_검사(): withMessage()를 통해 error message 일치 검사

YJGwon committed:
test: Set API에 대한 학습 Test class 추가

juniitstudy/SetTest.java 추가
   - set_size_확인(): set은 중복 제거를 해주기 때문에 중복 된 값 한번만 저장
   - set값_전체_존재_확인(int value): @ParameterizedTest와 @ValueSource를 이용해 여러개의 값을 파라미터에 대입해서 실행
   - set값_존재_여부_구별(int value, boolean result): @CsvSource를 이용해 두 파라미터에 여러쌍의 값을 대입, 자료형에 맞게 String값이 파싱된다

YJGwon committed:
feat: 문자열 계산기 최초 commit

YJGwon committed:
docs: stringcalculator\README.md 추가

기능 목록 최초 작성

YJGwon committed:
feat: 입력값이 null값 또는 빈 문자열인지 판별하여 0을 반환

StringCalculator.java
   - splitAndSum() 추가
   - castStringToInts(String input): String 판별해서 int Array로 cast
   - calculate(int[] numbers): cast된 배열의 총합을 계산

YJGwon committed:
docs: 완료된 기능 체크

YJGwon committed:
feat: 커스텀 구분자가 존재하는 경우

StringCaculator.java
   - separate(String input, String separator): 구분자로 문자열을 쪼개서 int array로 반환
   - getCustomSeparatorMatcher(String input): 커스텀 구분자 패턴 matcher 반환

YJGwon committed:
feat: 커스텀 구분자가 없는 경우

StringCalculator.java
   - castStringToInts(): : 과 , 를 구분자로 지정해서 separate() 호출

YJGwon committed:
feat: [예외] 변환한 숫자 배열 중 음수가 있는 경우

StringCalculator.java
   - validatePositive(): 숫자가 음수인 경우 RuntimeException 발생
   - separate(): Stream 중간 연산 메서드 peek을 사용해서 Stream에 영향을 주지 않고 모든 요소에 대해 validatePositive 실행
StringCalculatorTest.java
   - splitAndSum_not_number(): 숫자가 아닌 값일 경우에 Exception에 대한 테스트 미완성

YJGwon committed:
docs: README.md 기능 목록 최초 작성

YJGwon committed:
feat: 주어진 이름으로 Car 객체 생성

Car.java
   - private 멤버 name 추가
CarTest.java
   - 차_생성()

YJGwon committed:
feat: 입력값을 List<Car>로 만들어주는 CarFactory Class 추가

CarFactory.java : 정적 팩토리 패턴 응용

name에 대한 길이 검증 필요

YJGwon committed:
feat: [예외] 이름이 5글자를 초과할 때

Car.java
   - validateName(): 이름 5글자 이하인지 검사
CarTest.java
   - 이름_5글자_이상: assertThatThrownBy의 hasMessageStartingWith을 이용해 에러 메세지 비교

에러 메세지 출력과 재시작 구현 필요

YJGwon committed:
feat: [예외] 이름이 공백일 경우

Car.java
   - validateEmptyName(): 이름이 공백인지 검사
CarTest.java
   - 공백_이름()

에러 메세지 출력과 재시작 구현 필요

YJGwon committed:
feat: [예외] 차 이름이 중복인 경우

Car.java
   - equals(), hashCode(): Car 객체의 이름이 같은 경우는 같은 객체로 취급
CarFactory.java
   - of(): 리턴 타입을 List에서 Set으로 변경해서 collect하면서 바로 중복 제거
   - validateDuplicatedCarName(): 문자열 배열과 Set의 크기가 다를 때 중복 이름 에러 발생
CarTest.java
   - 중복된_이름(): 중복된 이름 에러 발생 테스트

YJGwon committed:
feat: 값이 4 이상일 경우 전진

Referee.java
   - judgeCarMove(Car car, int randomValue): 전달받은 값이 4이상일 경우 전진시킨다
Car.java
   - position: 차의 현재 위치를 나타내는 멤버 변수
   - move(): position에 1을 더한다
RefereeTest.java
   - 차_전진_판단: 4이상의 값을 줬을 때 차가 전진하는지 테스트

YJGwon committed:
feat: 난수 생성하기

Game.java
   - makeRandomValue(): 0부터 9사이의 랜덤 값 생성
GameTest.java
   - 랜덤값_부여_테스트(): 랜덤값이 범위에 맞게 생성하는지 검사

난수 생성은 굳이 테스트 하지 않아도 되지만, 범위에 대해 테스트 하는 좋은 방법을 고민해보기 위해 테스트 코드 작성

YJGwon committed:
feat: 우승자 판정

Referee.java
   - getMaxPosition(Set<Car> cars): stream의 max()를 이용해서 최대 position값 도출
   - judgeWinner(Set<Car> cars): getMaxPosition()에서 도출한 값과 position이 같은 Car 객체 List 반환
RefereeTest.java
   - 우승자_판정(): position값에 차이를 뒀을 때 우승자 판정 테스트
   - 우승자_여러명(): 우승자가 여러명일 때 테스트

YJGwon committed:
feat: InputView 추가

YJGwon committed:
feat: Car.toString 추가

CarTest.java
   - toString_kun(): 이름이 포함되는지
   - toString_forky(): 다른 이름을 넣어도 통과하는지(멤버변수 name 사용)
   - toString_position_1(): position charactor(-)가 포함되는지
   - toString_position_3(): position이 바뀌어도 통과하는지 (멤버변수 position 만큼 position charactor를 StringBuilder로 연결)

TDD 피드백 강의 이후 최소의 구현으로 test를 통과시키는 TDD Cycle을 적용해서 toString()을 완성

YJGwon committed:
feat: 이름, 횟수 입력받아서 게임 1회 진행

Game.java
   - start(): 실제 게임을 시작하고, 게임의 흐름을 제어하는 역할
   - play(): Car객체에 난수를 부여하고 전진하는 게임 진행 1회 수행
   - showResult(): 게임의 1회 진행 결과 출력

YJGwon committed:
feat: 우승자 출력

Game.java
   - showWinner(): Stream의Collectors.joining()을 사용해서 collect가 될 때원하는 형태의 문자열로 결합되게 함

YJGwon committed:
feat: 시도횟수만큼 반복

YJGwon committed:
feat: [예외] 횟수가 음수일 때

InputView.java 오류 fix
   - inputCarNames(): 이름에 공백이 있을 때 inputGameCount가 실행되며 발생하는 버그를 next() -> nextLine()으로 변경하여 해결

Hongdonggeon committed:
docs: README 파일 수정

Hongdonggeon committed:
refactor: 에러 발생 기준값 상수화

- Game.java
   입력하는 값이 0 이하일 때 예외를 발생하는데, 이 기준 값인 0을 상수화

Hongdonggeon committed:
refactor: 스트림 한 단계 중간 연산 추가

CarFactory.java
   -of(): 한번에 여러가지 작업을 하는 것보다 하나의 행위 위주로 한단계씩 작성하는 것이 좋다.

Hongdonggeon committed:
refactor: 클래스 내 메서드 위치 변화

Hongdonggeon committed:
refactor: position 초기값 0 명시

Car.java
   자동차 경주의 시작점이 바뀔 수 있고, 0에서 시작한다고 명시해주는 것이 더 명확함

Hongdonggeon committed:
refactor: Random 객체 멤버 변수에서 인스턴스 변수로 전달

Game.java
   랜덤값을 생성할 때 계속해서 객체를 만들어주는 것보다 Game 클래스가 만들어질 때 Random 객체 한번 생성

Hongdonggeon committed:
refactor : play() 메서드 반환형 void 로 변환

Game.java
   - play(): showResult(play(cars));로 바꿔줄 수 도 있었지만 지금 처럼 작성한 것이 가독성도 더 좋고 자동차 경주 게임 흐름을 더 잘 표현한다고 생각했다.

Hongdonggeon committed:
refactor: 예외 발생 테스트 코드 에러문 비교하는 메서드 변경

CarTest.java
   - hasMessage(): 발생한 에러의 에러문과 정확히 문자열이 일치하는지 비교
   - hasMessageContaining(): 발생하는 에러문에 파라미터로 들어온 문자열이 포함되는지
   - hasMessageStartsWith() : 발생하는 에러문과 파라미터로 들어온 문자열 처음부터 비교, 정확히 문자열이 일치하지 않아도 되고 초반부터 똑같기만 하면 됨

Hongdonggeon committed:
refactor: judgeCarMove 역할 분리

- Referee.java
  - canCarMove():  이전에는 Referee 객체가 랜덤값을 부여받아 Car객체를 움직여 주었다면, Referee가 랜덤값을 부여받아 Car가 움직일 수 있는지 없는지 판정만 내려준다.
Car.java
   - move(boolean): 이전에는 단지 움직이는 메서드였다면 Referee 객체에게 움직일수있다는 판정을 받으면 움직이기

이외의 변화된 클래스들은 메서드의 구조가 바뀌면서 자연스럽게 바뀌게 되었다.

Hongdonggeon committed:
Test: 랜덤값에 따른 전진여부 판단 테스트 추가

RefereeTest.랜덤값이_3일때_차_전진_판단
RefereeTest.랜덤값이_4일때_차_전진_판단
RefereeTest.랜덤값이_5일때_차_전진_판단
   - 테스트를 작성할 때 기준이 되는 값 전, 후를 다 테스트 해주는것이 좋다.

Hongdonggeon committed:
Test: Car객체 생성 테스트

CarTest.이름_4글자_일때_차_생성
CarTest.이름_5글자_일때_차_생성
CarTest.이름_6글자_일때_차_생성
   - Car 객체의 이름의 최대 길이 기준값인 5와 전 후 테스트

Hongdonggeon committed:
feat: 여러 Car객체를 관리하는 Cars 클래스 추가

Cars.java
   Car객체들을 관리하는 일급 컬렉션, 생성자에 List<Car>를 파라미터로 받아 객체 생성

CarsTest.java
   여러개의 Car객체들이 빠짐없이 Cars 객체의 Set<Car>에 추가되는지 테스트

Hongdonggeon committed:
refactor: Car 객체 생성해주는 팩토리 메서드 역할 분리

CarFactory.java
   of(): 이전에는 문자열을 쪼개서 중복 검사를 해주고 Set으로 반환했다면, 증복검사는 Cars 객체에 위임하고 단지 문자열을 쪼개서 Car객체만 생성해주는 기능만 하도록함

Hongdonggeon committed:
feat: Cars 객체를 만들 때, 중복 Car 객체 검사해주기

Cars.java
   validationDuplicatedCarName(): 스트림의 distinct()를 사용해서 중복 검사해주기
   getCars(): unmodifiableSet()을 이용해서 불변 객체 반환해주기
CarsTest.java
    중복된_이름(): CarTest에서 CarsTest로 이동
CarTest.java
   차_1대_생성(): Car 객체를 만들어주는 CarFactory.of()에 대한 테스트 추가

이외의 변화들은 여러 Car 객체를 관리하고 반환해주는 Cars 클래스를 추가함에 따라 생긴 자연스러운 변화들임

Hongdonggeon committed:
refactor: Car 객체 집합에서 최대값 position 찾는 메서드 위치 이동

Cars.java
   getMaxPosition(): 이전에는 Referee 클래스에 존재하였는데 Set<Car>를 관리해주는 Cars 객체에 역할 위임

Referee.java
   judgeWinner(Set<Car>, int): Referee는 최대값 position을 파라미터로 받아 파라미터로 받는 Set<Car>에서 최대 position 값 과 position이 똑같은 Car 들만 필터링해줌

Game.java
   Cars의 getCars()를 계속해서 호출한 것보다, 한번만 호출하여 사용하는게 더 낫다고 생각

이외의 변화는 메서드 구조의 변화로 자연스럽게 바뀜

Hongdonggeon committed:
feat: Cars의 Set<Car>에서 이름이 같은 Car 반환

Cars.java
   getCarByName(String): Cars의 Set<Car>에서 파라미터로 들어오는 문자열과 이름이 같은 Car 객체 반환

CarsTest.java
   getCarByName()을 이용해 이름이 같은 Car가 잘 반환되는지 테스트

Hongdonggeon committed:
Test: Cars의 getCarByName()과 getMaxPosition() 테스트 코드 추가

CarsTest.Car_집합에서_찾는_이름이_존재_X(): 찾는 이름이 존재하지 않은 경우

Hongdonggeon committed:
refactor: 사용하지 않은 상수 삭제

Hongdonggeon committed:
refactor: 메서드 체이닝 한줄에 한개씩 하기

Hongdonggeon committed:
Test: 프로덕션 코드에서 사용하지 않는 메서드를 사용하지 않고 테스트 코드 수정

Cars.java
   - getCarByName() : 프로덕션 코드에서 사용하지 않으므로 메서드 삭제

CarsTest.java
   - Car_집합에서_최대_position_값_찾기() : getCarByName()을 사용하지 않고 테스트 코드 수정

Hongdonggeon committed:
test: test 메서드 given when then 방식으로 구성

Hongdonggeon committed:
refactor: Cars 일급 컬렉션 객체 List로 변경

Set은 순서를 보장해주지 않기 때문에, 자동차 경주 결과값을 출력할 때 입력한 순서대로 결과가 나오지 않았다. List로 변경하여 순서 보장

Hongdonggeon committed:
refactor: Cars List<Car>에서 position이 같은 Car 객체들 반환해주는 메서드

이전에는 Referee의 judgeWinners()에서 position이 같은 Car들을 반환해주는 기능을 구현했었는데, 만약 꼴등도 뽑아야 하는 기능도 추가가 된다면 judgeLosers()라는 메서드를 또 구현해주어야 한다. 중복된 코드가 많아지므로 일급 컬렉션에게 행위를 위임하고 행위를 요청하는 방식으로 리팩토링

Hongdonggeon committed:
feat: 숫자를 반환해주는 인터페이스 추가

인터페이스
NumberGenerator.java : 숫자를 반환해주는 인터페이스

구현체
RandomNumberGenerator.java : 0 ~ 9의 랜덤 숫자를 반환해준다. (프로덕션 코드에서 사용)

MovableNumberGenerator.java : 4를 반환해준다. 차가 전진하는지 검증하는 테스트에서 사용 (테스트 코드에서 사용)
NonMovableNumberGenerator.java : 3을 반환해준다. 차가 전진하지 못하는 검증하는 테스트에서 사용 (테스트 코드에서 사용)

Hongdonggeon committed:
feat: 뷰에 데이터를 전달하기 위한 CarDto 클래스 추가

Car.java
  toString()은 객체를 표현하는 메서드로 객체를 표현해야 하는 상황이 오면 사용하지 못하게 된다. 그리고 뷰가 해야하는 일을 Car가 해버린 것 일수도 있다.

OutputView.java
  뷰에게 데이터만 전달하고 뷰가 그 데이터로 UI를 생성하게 변경

Cars.java
  toCarDto() : List<Car>에서 List<CarDto>로 다른 형태의 겍체로 변환해주기

Hongdonggeon committed:
feat: domain, view, controller, dto 로 패키지 구성

domain
   - car
      - Car.java
      - CarFactory.java
      - Cars.java

   - numberGenerator
      - NumberGenerator.java
      - RandomNumberGenerator.java
      - MovableNumberGenerator.java
      - NonMovableNumberGenerator.java

   - Referee.java

view
   - InputView.java
   - OutputView.java

controller
   - Game.java

dto
   - CarDto.java

Hongdonggeon committed:
test: 게임시도횟수 음수 입력한 경우 예외발생 테스트, 게임 시도 횟수 제대로 입력한 경우 인스턴스 생성 테스트

Hongdonggeon committed:
feat: Game과 GameController 역할 분리

분리 하기 전, GameController가 했던 일
 - 시도 횟수 검증
 - start() : 게임 시작하기 - 자동차 경주 게임 전체 플로우 조립
 - play() : 랜덤값에 따라 Car 움직이게 하기
 - showResult() : 실행 결과 보여주기
 - showWinner() : 우승자 보여주기

 Controller 에는 뷰와 비즈니스 로직을 합쳐주는 것 이외에 역할을 Game로 분리

Game.java
- 게임 시도 횟수 검증
- play()
그리고 Referee 우승자 판정도 Game에게 역할 위임
Game은 자동차 경주 게임에 있어서 핵심 기능을 하는 클래스

Hongdonggeon committed:
refactor: 우승자 출력문 만들기 기능 역할 위임

이전에는 Controller에서 스트림 joining을 이용해서 문자열을 OutputView 한테 보내줬는데,
이 출력문 만드는 기능은 View 에 있는 것이 더 적합하다고 생각해서 Controller에서는 WinnerCarDto를 전달해주고 View가 전달 받은 값으로 우승자 출력문 생성

Hongdonggeon committed:
test : car 리스트에서 position이 같은 Car 찾기 테스트코드 추가

Hongdonggeon committed:
style: 코드 포매팅

Hongdonggeon committed:
Merge branch 'hongdonggeon' into step2

Hongdonggeon commented on src/main/java/racingcar/controller/GameController.java:
```diff
@@ -0,0 +1,44 @@
+    private final Game game;
+
+    public GameController() {
+        game = new Game(CarFactory.of(InputView.inputCarNames()), InputView.inputGameCount());
```
기본 생성자를 이대로 사용하는 것은 안 좋은 방법일까요?

lowoon replied:
이대로 사용한다는 것이 어떤걸 말하는 걸까요?? 생성자 내에서 new 선언을 말하는 걸까요??

다른 내용으로 view가 controller까지 넘어와도 괜찮을까요??
참고하시면 좋을거 같아요. https://tecoble.techcourse.co.kr/post/2021-04-25-dto-layer-scope/

Hongdonggeon replied:
Controller에서 View를 넘기지 않고 데이터를 전달하는 방식으로 리팩토링 해보았습니다!!

Hongdonggeon commented on src/main/java/racingcar/controller/GameController.java:
```diff
@@ -0,0 +1,44 @@
+
+    public void showWinner() {
+        List<WinnerCarDto> winnerCarDtos = game.getCarWinnerDto();
+        OutputView.printWinner(winnerCarDtos);
```
우승자를 출력하는 기능을 위해, View에 Dto를 넘겨줄 때, CarDto를 사용을 하려다 CarDto에는 우승자 출력에 필요없는 position도 있기 때문에 사용하지 않고 WinnerCarDto를 따로 만들었어요!! 
CarDto에는 position과 name, WinnerCarDto에는 name만 멤버 변수로 존재하는데, 보내야할 데이터를 포함하고 있는 Dto가 이미 존재한다면 이전에 만들어 놓은 Dto를 그냥 사용을 해도 될까요?

lowoon replied:
**개인적으로는** 사용해도 상관 없다고 생각합니다.
새로 만드는 것이 명확하고 불필요한 필드가 없어 좋으나 일을 하다보면 dto객체가 생각보다 많이 생겨 관리가 힘들 때가 있어서요...
List전체를 감싸고 각 라운드의 결과와 우승자를 모두 가진 dto를 만들수도 있지 않을까요??

Hongdonggeon replied:
답변 감사드립니다!! 지금은 Dto를 별로 안 만들어도 되지만, 큰 프로그램에서는 Dto가 많이 생기게 되면 그만큼 관리도 어려워지겠네요!! 그리고 View에 List<Dto>로 보내지 않고 Dto객체 하나만 보내주기 위해, Dto안에 멤버 변수를 List로 다 변경시켰습니다!!

Hongdonggeon commented on src/main/java/racingcar/controller/GameController.java:
```diff
@@ -0,0 +1,44 @@
+
+public class GameController {
+
+    private final Game game;
```
현재 game이 멤버 변수로 존재하긴 하는데, GameController의 메서드들이 인스턴스의 상태를 변경하거나 반환하는 일을 하는 것은 아닌 것 같아서요,,, 모든 메서드를 static 메서드로 바꿀까 고민을 했습니다. GameController 같이 모든 코드를 조립해주는 클래스들을 어떻게 해야 더 객체지향적인 코드로 설계할 수 있을지 궁금합니다!

lowoon replied:
이런 방식으로도 생각해볼 수 있을거 같은데요. game이 하는 것을 controller에서 할 수도 있지 않을까요?? 
controller의 역할이 무엇인가요?? controller의 역할에 대해 고민해보는게 좋을거 같아요.

Hongdonggeon replied:
View에서 데이터를 전달받고 데이터를 전달해주는 방식으로 다시 리팩토링 해보니까 Controller의 역할에 대해 더 감이 왔던 것 같아요!! View와 데이터를 주고 받으며 Car 객체들을 만들고 움직이게하고 우승자를 판정하는 등의 비즈니스 로직들을 실행시키는 역할을 하는 것 같아😀

lowoon replied:
네 맞습니다!
추후에 프로그램이 커지면 service가 생기면서 비지니스 로직이 분리가 될거에요!
분리를 할때 지금의 고민이 도움이 될거라고 생각합니다 👍

Hongdonggeon commented on src/main/java/racingcar/domain/Game.java:
```diff
@@ -0,0 +1,55 @@
+        return judgeWinner().stream()
+                .map(WinnerCarDto::new)
+                .collect(Collectors.toList());
+    }
```
Game 클래스에 List<Car>에서 List<Dto>로 변환을 해주는 메서드들이 있어도 될까? 라는 고민을 많이 했던 것 같아요... 뭔가 이렇게 형변환 해주는 메서들을 어디에 두어야 될까요? Game에 추가한 이유로는, GameController의 멤버 필드인 Game을 최대한 잘 활용하기 위해서 입니다!

lowoon replied:
game에서 dto를 만들어주는 메서드를 만들게 되면 dto가 추가될 때마다 메서드가 추가될 텐데요.
dto에서 car를 통해 자기 자신을 만들 수 있도록 할 수도 있을거 같습니다. 생성의 책임을 어디에 둘건지에 따라 다를거 같아요.

Hongdonggeon replied:
넵!! Dto가 추가될 때마다, Game에 추가하는 것은 안 좋을 것 같다는 생각이 드네요!! Dto로 변환해주는 메서드들은 Dto 객체에 있는 것이 어울리다고 생각해서, 팩토리 메서드 패턴을 활용하여 해당하는 Dto 객체에 위치시켰습니다

lowoon commented on src/main/java/racingcar/domain/Game.java:
```diff
@@ -0,0 +1,55 @@
+        for (Car car : cars.getCars()) {
+            car.move(numberGenerator);
+        }
+        return cars;
```
```suggestion
    public Cars play(NumberGenerator numberGenerator) {
        cars.move(numberGenerator);
    }
```
형태로 만들수도 있지 않을까요??

Hongdonggeon replied:
너무 너무 좋은 방법인 것 같아요! Cars에 move라는 메서드를 만들어주면서 getter의 사용도 줄였고 테스트 코드 작성하기도 더 쉬워졌네요!! 일급 컬렉션을 만들면, 잘 활용하는 방법에 대해 항상 생각해봐야겠습니다!!

lowoon commented on src/main/java/racingcar/domain/car/CarFactory.java:
```diff
@@ -0,0 +1,16 @@
+
+public class CarFactory {
+
+    public static List<Car> of(String names) {
```
- from : 매개변수를 하나 받아서 해당 타입의 인스턴스를 반환하는 형변환 메서드
- of : 여러 매개변수를 받아 적합한 타입의 인스턴스를 반환하는 집계 메서드

Hongdonggeon replied:
아~ 그렇군요! 감사합니다!! 리뷰 반영했습니다 :)
여기 메서드에서는 createCars라고  메서드 이름을 짓는 것보다, from으로 짓는게 더 좋을 까요?
의미가 너무 모호해지진 않을까라는 생각이 들어서요! 

또한 Dto로 변환해주는 메서드들도 from()으로 바꿔주려고 하는데, from이라는 메서드가 너무 많아지면 헷갈리지 않을까 싶어서요,,! 
반대로 생각해서, from()이라고 명명 함으로써, 다른 사람들이 봤을 때, 아 형변환 해주는 메서드구나 하고 더 의미가 확실해 질까요?

lowoon replied:
말씀하신 from의 경우처럼 자기 자신을 반환하는 경우에는 의미가 확실한데, Factory처럼 자기 자신이 아닌 다른 것을 생성해 주는 것은 createCars처럼 메서드를 통해 생성해주는 객체를 나타내주는게 좋다고 생각합니다~

lowoon commented on src/main/java/racingcar/domain/car/Cars.java:
```diff
@@ -0,0 +1,53 @@
+    private static final String DUPLICATED_NAME_ERROR = "[ERROR] 차 이름은 중복될 수 없습니다.";
+    private static final String NOT_EXIST_CAR = "[ERROR] 차가 존재하지 않습니다.";
+
+    private final List<Car> cars;
```
단순히 순서때문이였으면 LinkedHashSet 도 있습니다~

Hongdonggeon replied:
와,, 그런 방법이 있었네요! 갓자바다 😀

lowoon commented on src/main/java/racingcar/domain/car/Car.java:
```diff
@@ -0,0 +1,69 @@
+    private static final int MOVE_CRITERIA = 4;
+
+    private final String name;
+    private int position;
```
너무 잘해주셔서 한가지 도전해보죠 😎 
name과 position 각각을 감싼 객체를 만들어 볼까요??

Hongdonggeon replied:
넵!! 원시값 포장!! Name과 Position이라는 객체를 만들어줬어요!

lowoon reviewed (changes requested):
반가워요 쿤! 2단계도 잘 부탁해요 😄
1단계 마무리에 남김 피드백이 다 반영이 되었네요 👍
질문하셨던 부분에 답변달았고, 피드백 몇 가지 남겨봤으니 확인 부탁드려요.

궁금한 점이 생기면 언제든지 DM 혹은 댓글남주세요 ^-^

Hongdonggeon committed:
refactor : Car 객체들을 생성해주는 CarFactory 메서드 of에서 from 으로 이름 변환

- of() : 여러 매개변수를 받아 적합한 타입의 인스턴스를 반환하는 집계 메서드
- from() : 매개변수를 하나 받아서 해당 타입의 인스턴스를 반환하는 형변환 메서드

Hongdonggeon committed:
Merge branch 'step2' of https://github.com/Hongdonggeon/java-racingcar-1 into step2

Hongdonggeon committed:
중복 파일 삭제

Hongdonggeon committed:
feat : Cars 객체에 차들을 움직이게 하는 메서드 추가

이전에는
Game의 멤버 필드인 cars.getCars()로 List를 받아와서 for문으로 Car하나씩 전진시켜줬는데,
Cars 객체에 move라는 메서드를 만들어주면서 getter의 사용을 줄일 수 있었고, 일급 컬렉션 객체를 더 잘 활용하게 되었다.

Hongdonggeon committed:
feat : Dto 구조 변경, Dto 변환 메서드 위치 변경

CarDto.java -> CarsDto.java
 Car 하나의 정보만 가지고 있었는데, List<Car>의 정보를 가지고 있다.

WinnerCarDto.java -> WinnerCarsDto.java
 WinnerCar 하나의 정보만 가지고 있었는데, 전체 WinnerCar의 정보를 가지고 있다.

팩토리 메서드 패턴을 이용해, Dto 변환 메서드를 해당하는 Dto 클래스에 생성

Hongdonggeon committed:
feat : Car 객체 이름 원시값 포장

Name 포장 객체에서 이름 규칙에 맞는 검증을 해준다.
car.getName.getName() 처럼 무분별한 getter 사용을 하지 않기 위해, Car의 Name 객체를 활용해 메서드를 만들어준다

Hongdonggeon committed:
feat : Car 객체 포지션 원시값 포장

Hongdonggeon committed:
feat : Controller 역할 재분배 및 분리

이전에는 Controller에서 바로 View를 사용했었는데, InputView의 입력 데이터를 Controller 메서드에서 파라미터로 전달받고, OutputView에서 필요한 데이터 또한 Dto로 전달

createCars : Car 객체들 생성
createGameCount : 게임 시도횟수 검증 후 반환
move () : Car객체들을 1라운드 실행하고 Dto로 결과값 전달
judgeWinner() : 우승자 판정 후 Dto로 결과값 전달

Hongdonggeon committed:
feat: Cars객체에 우승자 판정 메서드 추가

반환값을 WinnersDto로 선언하여 컨트롤러에서 Dto로 변환해주지 않고 서비스단에서 반환해 컨트롤러에서 그대로 사용한다.

Hongdonggeon committed:
style : 코드 포맷팅

lowoon commented on src/main/java/racingcar/domain/car/Position.java:
```diff
@@ -0,0 +1,21 @@
+package racingcar.domain.car;
+
+public class Position {
```
참고용이에요.
Integer를 감싼 객체라서 Comparable을 적용해서 Position 자체가 비교되도록 할 수도 있습니다~

Hongdonggeon replied:
Position에서 Comparable 인터페이스 구현하여 Car 에서 사용한 코드인데요,,
```java
    public boolean isPosition(int positionValue) {
        return position.compareTo(new Position(positionValue)) == 0;
    }
```
isPosition()메서드를 사용할 때마다, 새로운 Position 객체가 생성 되는 것이 조금 걸리는데,,, 괜찮을까요?

Hongdonggeon replied:
다른 생각이 나서 리팩토링 해주었어요!! 
```java
    public boolean isPosition(Position position) {
        return this.position.compareTo(position) == 0;
    }
```

```java
    public Position getMaxPosition() {
        return cars.stream()
                .map(Car::getPosition)
                .max(Comparator.comparing(position -> position))
                .orElseThrow(() -> new IllegalArgumentException(NOT_EXIST_CAR));
    }
```
그리고 테스트 코드에서도 Position 비교할 때, 이렇게 비교할 수 있더라구요!!
```java
assertThat(cars.getMaxPosition()).isEqualByComparingTo(new Position(1));
```
이전에 getPosition()이 Position객체의 getPositon() 메서드를 한번 더 호출하는 것이었는데, Position 자체로 비교를 하니 메서드도 덜 호출되고, 코드도 더 깔끔해진 것 같아요!!
마지막까지 정말 새로운 방법을 너무 많이 배워가네요!! 너무 감사드립니다 😀

lowoon commented on src/main/java/racingcar/dto/CarsDto.java:
```diff
@@ -0,0 +1,34 @@
+    private List<Integer> carPositions;
+
+    public CarsDto(List<String> carNames, List<Integer> carPositions) {
+        this.carNames = carNames;
```
2개의 List를 가지고 index를 통해 값을 보여주는데 Map으로 이름과 위치가 매칭되도록 저장할 수 있을거 같아요.
carsDto를 말씀드린 이유는 게임의 결과 데이터 이외에 다른 날짜, 시간, message등 다른 데이터가 추가될 때 용이할 수 있어서 말씀드렸었습니다~

Hongdonggeon replied:
Map으로 바꾸니까 더 깔끔해졌어요!! View에서 출력할 때도 forEach문으로 하니 더 깔끔해진 것 같구요!! 
아 그럴 수 있네요 오로지 게임 결과이외에도 다른 데이터가 추가 될 수 있네요!

lowoon reviewed (approved):
안녕하세요 쿤!😄
질문하면서 학습하고, 점점 더 코드가 변화하는 모습을 보면서 저도 기분이 좋아졌던거 같아요 👍
잘 따라와 주셔서 저도 욕심이 나서 여러 부분을 도전했던거 같네요.
간단한 코멘트만 남겼으니 확인 부탁드려요.
미션은 끝났지만 궁금한점 있으면 DM이나 코멘트로 남겨주세요!

1, 2단계 자동차 경주 미션 수고하셨습니다!
